### PR TITLE
Fix bugs in TimeDiscretization and enhance TimeDiscretizationInterface

### PR DESCRIPTION
**Pull request by delreluca**
See finmath-lib issue #58

We unify duplicate removal and tick size rounding in TimeDiscretization
Building upon that we define unions and intersections of discretizations

### What is this PR for?
Unifying the handling of tick sizes and duplicates in the `TimeDiscretization` class in order to add unions and intersections to the `TimeDiscretizationInterface`.

### What type of PR is it?
Bug Fix, Refactor and Feature

### Todos
* I would suggest revisiting retrieving the default tick size from a system property, as now it can be set when constructing the `TimeDiscretization`. It makes the thing harder to test and it might clash when two dependencies both use finmath with different tick sizes.

### What is the related issue?
https://github.com/finmath/finmath-lib/issues/58

### How should this be tested?
Unit tests are provided

### Questions:

**Does the licenses files need update?**

NO
 
**Are there breaking changes for older versions?**

YES
If there was any reliance on duplicates or unrounded time points. I ran the whole unit test suite and it didn’t seem to be the case there.

**Does the change require additional documentation?**

NO

### CODE REVIEW

**Comment by cfries** [comment]
I had a very quick look, but looks to me as if the old constructor did a sort, but the new one does not. This might be an issue. Will check asap.

File: src/main/java/net/finmath/time/TimeDiscretization.java [hunk 1]
```diff
@@ -29,71 +30,99 @@
...
 
+    /**
+     * Constructs a time discretization using the given tick size.
+     *
+     * @param times    A non closed and not necessarily sorted stream containing the time points.
+     * @param tickSize A non-negative double representing the smallest time span distinguishable.
+     */
+    public TimeDiscretization(DoubleStream times, double tickSize) {
+        this.timeTickSize = tickSize;
+        this.timeDiscretization = times.map(this::roundToTimeTickSize).distinct().sorted().toArray();
```
Comment (delreluca) [inline src/main/java/net/finmath/time/TimeDiscretization.java]: @cfries This line takes care of the sorting. The nice thing is that all other constructors will eventually end in this constructor.

Comment (cfries) [reply src/main/java/net/finmath/time/TimeDiscretization.java]: Beautiful! I will have look. (The travis fail was my fault).

File: src/main/java/net/finmath/time/TimeDiscretization.java [hunk 2]
```diff
@@ -10,6 +10,7 @@
 import java.util.Arrays;
 import java.util.Iterator;
 import java.util.Set;
+import java.util.stream.*;
```
Comment (cfries) [inline src/main/java/net/finmath/time/TimeDiscretization.java]: No star imports please ;-)

File: src/main/java/net/finmath/time/TimeDiscretizationInterface.java [hunk 3]
```diff
@@ -90,4 +90,14 @@
 	 * @return A new time discretization where all time points have been shifted by the given time shift.
 	 */
 	TimeDiscretizationInterface getTimeShiftedTimeDiscretization(double timeShift);
+
+	/**
+	 * @param that Another time discretization containing points to add to the time discretization.
```
Comment (cfries) [inline src/main/java/net/finmath/time/TimeDiscretizationInterface.java]: If JavaDoc is present a description should be present. The behaviour of tickSize within the union and intersection should be documented. To be honest: The tickSize feature (there was a use case for this) makes me wonder if union and interception are well defined operators. For example we have hat for two disjoint sets A and B (A union B) interesect B is not equal to B.  - not sure here. API has to be clarified (max/min on tickSize?!)

File: src/main/java/net/finmath/time/TimeDiscretization.java [hunk 4]
```diff
@@ -204,15 +218,39 @@ public int getTimeIndexNearestGreaterOrEqual(double time) {
...
+
+    /**
+     * @param that Another time discretization containing points to add to the time discretization.
+     * @return A new time discretization containing both the time points of this and the other discretization.
+     */
+    @Override
+    public TimeDiscretizationInterface union(TimeDiscretizationInterface that) {
+        return new TimeDiscretization(
+                Stream.concat(Arrays.stream(timeDiscretization).boxed(), Arrays.stream(that.getAsDoubleArray()).boxed()),
+                Math.max(timeTickSize, that.getTickSize()));
```
Comment (cfries) [inline src/main/java/net/finmath/time/TimeDiscretization.java]: Why max? Better min? If max: two points which where different in A are no longer different in A union B, even if B is empty. If A is "rounded to daily" and B is "rounded to monthly" the union should be "rounded to daily".

Comment (cfries) [reply src/main/java/net/finmath/time/TimeDiscretization.java]: Not sure. Maybe like that: for intersect it is max. For union it is min. In both cases we have to have tickSize1 = integer * tickSize2

File: src/main/java/net/finmath/time/TimeDiscretization.java [hunk 5]
```diff
@@ -29,71 +30,99 @@
...
 	}
 
+    /**
+     * Constructs a time discretization using the given tick size.
+     *
+     * @param times    A non closed and not necessarily sorted stream containing the time points.
+     * @param tickSize A non-negative double representing the smallest time span distinguishable.
+     */
+    public TimeDiscretization(DoubleStream times, double tickSize) {
+        this.timeTickSize = tickSize;
```
[the remark this answers is not in the document]
Comment (delreluca) [reply src/main/java/net/finmath/time/TimeDiscretization.java]: I don’t disagree, my IDE messed this file up I guess

**Comment by delreluca** [comment]
I will adjust the design changes later this day (asterisk import, tabs, missing Javadoc).

Regarding the tick sizes: I agree that it should be min for unions and max for intersections. I’m not sure how to reliably enforce that the tick sizes are integer multiples of each other for doubles, I guess I will not check and warn about it in the Javadoc like this: “If the tick sizes are not integer multiples of each other multiple operations might alter the time points, for example a.intersect(B.union(a)) might be different from a“.

What do you think about it?

Another possibility would be to remember the unrounded time points, but I guess at that point the user could remember it themselves and re-create a TimeDiscretization with a common tick size.

**Comment by cfries** [comment]
Your solution is fine. Remembering the original points sound good, but looks wasteful to me… lets don’t do it.
What do you thing about requiring tick sizes to be equal? The tick side is like a unit and you can union „days with days“ and „years with years“ but you are not allow to mix. (I would bet that we do not have the use case of different tick size, because here the tick size is just to have a robust definition of „equal“ the floating point times.

**Comment by delreluca** [comment]
I think requiring the same tick size too restrictive. In many cases users
will not bother setting a non-default tick size. But if they do, they will
have to research these edge cases anyway, and they will find the answer in
the Javadoc.

Here is an example. Someone might reduce the tick size to do an intraday
volatility simulation with GARCH, but they want to use the simulated
volatilities in a Euler scheme with a coarser time grid where they don’t
specify a tick size. (Somewhat contrived example I admit)

**Comment by cfries** [comment]
Another thought on this:

Since tickSize is used to define the equivalent classes of „times“ I was thinking about the need to change it. Maybe  someone going intraday and someone going for end of day valuation would both be satisfied with „times are equal if they are equal up to the milli-second. That is tickSize is 1.0 / (365.0 * 24.0 * 60.0 * 60.0 * 1000.0). This is 3E10, leaving 1E5 years as the the highest possible time horizon. But since we are using IEEE754, best would be to have tickSize being a rounding of the mantissa only (which could be 10^-12 then). …

**Comment by delreluca** [comment]
Hope the second commit resolves the issues.

Regarding changing the default to 1 millisecond, we could do that, but I guess that would be a different issue. By the way I’d also reconsider getting the default from the system property, as this seems a bad practice if the library is used as a dependency by two clients.

**Comment by cfries** [comment]
The properties file is an integral part of the distributed JA and a user has to actively overwrite that property. Why bad practice?

**Comment by delreluca** [comment]
Consider a Java program that depends on two other Java libraries A and B each depending on finmath.  They run on the same JVM. A sets the default tick size to 1 ms, but B sets it to 10 ms. Now the default tick size depends on the order A’s and B’s setProperty calls are executed.

**Comment by cfries** [comment]
I see the point.

**Comment by cfries** [comment]
I would favour a hard coded default, now that tickSize can be changed at construction.

**Comment by delreluca** [comment]
Me too, it fits with the new constructors. I’m not sure whether we should do it in this PR, though. If we change that line we might as well also change the default value itself as already suggested.

What’s your take on how changing the default (to 1 ms for example) might break existing code?

**Comment by cfries** [comment]
With respect to changing the default: My rule is that for revision a.b.c a change in c does not break code, a change in b does not break interfaces, but may lead to changes in behaviour and a change in a has massive changes in interfaces.  Changing the default is fine, but I would do a 4.3.0 release then. (Should be in a separate PR).  (Note: I have violated my own rule, but I try comply with it in the future).

**Pull request merged**